Changes to support Telco requirements

abhis2112 opened the pull request:
Changes:
1. Show interface name as labels for node to leaf edges.
2. Show multiple links for multiple ifaces.
3. Remove VM_Host from graph(Temp change for baremetal case)
4. See the interface details on the pod-to-node graph.
5. Color code edge for primary iface link and links of secondary ifaces.
6. Read the nodefabricnetworkattachment, get connection info and provide the new link for worker node to leaf node (different coloring).
7. Remove arrows from "Pod & Nodes" view.
8. Fix bug where the pod to node edges were duplicated on graph regeneration.
9. Read the pod based interface information from nodefabricnetworkattachment CR.
10. Changes for modified structure of nodefabricnetworkattachment CR.
11. Move role and rolebinding to access nodefabricnetworkattachments from neo4j template to vkaci template.
12. Modify the queries as per the new pod information from nodefabricnetworkattachment CR.
13. Add PF, VF for SRIOV ifaces.
14. Fix bug where duplicate nodes were shown.
15.  Fixed query incase node_leaf_all_iface_conn is empty to show worker node and its pods
16.  Removed the VM host between the node(s) and leaf(s) for baremetal case.
17.  Fixed the bug of links between node(s) and leaf(s) getting same label for baremetal case
18.  Fixed the issue of graph not updating on regenerating the graph after deleting the CR.
19. Removed labels from the queries
20. Wrapped the CR fetch in try except block
21. Removed dragview from neovis
22. Removed unnecessary commas while showing interfaces name
23. Added optional match for primary link query
24. Fix node/pod topology in absence of VM_HOST node
25. Add auto-detect functionality for ESX/Baremetal
26. Fix labels for the node
27. Added more UTs related to SRIOV and macvlan
28. Check nfad type based on primarycni field
29. Adding check in helm template to add role/rolebinding if CNO is enabled

abhis2112 committed:
Changes to support Telco requirements

Changes:
1. Show interface name as labels for node to leaf edges.
2. Show multiple links for multiple ifaces.
3. Remove VM_Host from graph(Temp change for baremetal case)
4. See the interface details on the pod-to-node graph.
5. Color code edge for primary iface link and links of secondary ifaces.
6. Read the nodefabricnetworkattachment, get connection info and provide the new link for worker node to leaf node (different coloring).
7. Remove arrows from "Pod & Nodes" view.
8. Fix bug where the pod to node edges were duplicated on graph
   regeneration.
9. Read the pod based interface information from nodefabricnetworkattachment CR.
10. Changes for modified structure of nodefabricnetworkattachment CR.
11. Move role and rolebinding to access nodefabricnetworkattachments from
   neo4j template to vkaci template.
12. Modify the queries as per the new pod information from
   nodefabricnetworkattachment CR.
13. Add PF, VF for SRIOV ifaces.
14. Fix bug where duplicate nodes were shown.
15.  Fixed query incase node_leaf_all_iface_conn is empty to show worker
    node and its pods
16.  Removed the VM host between the node(s) and leaf(s) for baremetal
    case.
17.  Fixed the bug of links between node(s) and leaf(s) getting same
    label for baremetal case
18.  Fixed the issue of graph not updating on regenerating the
    graph after deleting the CR.
19. Removed labels from the queries
20. Wrapped the CR fetch in try except block
21. Removed dragview from neovis
22. Removed unnecessary commas while showing interfaces name
23. Added optional match for primary link query
24. Fix node/pod topology in absence of VM_HOST node

abhis2112 commented:
Work in Progress. Do not merge. Currently for tracking changes

burhan20 committed:
Telco requirements:

Changes:
1. Removed arrows from pod, leaf and node topology
2. Added filter for bgp peers in table view
3. Added VM in between node and leaf for SRIOV/MacVlan
4. Modified the logic and wrote additional queries to support
    baremetal, VM host, and a combination of both
5. Fixed crash in Node and Pod topology for non baremetal case
6. Fixed crash in Node Topology in upstream code when node is not
    connected to leaf or VM host

burhan20 committed:
Optimized and reordered queries, modified info on hover over node, added info to VM

burhan20 committed:
Displayed primary iface on hovering over node in non baremetal case and preceeded iface name with leaf name

abhis2112 committed:
Merge pull request #5 from noironetworks/burhan-telco

1. Support Hybrid Topology
2. Adding comments to queries
3. Correct Information when hovering node and VM_HOST

burhan20 committed:
Showing sec links only when pod is also present and removed extra iface from primary link

abhis2112 committed:
Merge pull request #8 from noironetworks/vkaci-comments

Showing sec links only when pod is also present and removed extra ifa…

burhan20 committed:
fixed ut failure

abhis2112 committed:
Merge pull request #9 from noironetworks/ut-failure

fixed ut failure

burhan20 committed:
Added more testcases for sriov and macvlan

burhan20 committed:
Made changes to queries so that when pod is present then only node to leaf/vm links are displayed in primary, sriov and macvlan cases

burhan20 committed:
Fixed Pod and Node topology view so that node-leaf/vm links are visible only when pod is also present

abhis2112 committed:
Merge pull request #11 from noironetworks/sriov-macvlan-pod

Made changes to queries so that when pod is present then only node to…

burhan20 committed:
optimized testcases

burhan20 committed:
fixed query crash if there is no pod with primary link running on a node

abhis2112 committed:
Merge pull request #10 from noironetworks/testcases

Added more testcases for sriov and macvlan

abhis2112 committed:
Merge pull request #12 from noironetworks/node-with-only-sec-pods

fixed query crash if there is no pod with primary link running on a node

burhan20 committed:
added ut for sriov macvlan pod in diff ns

abhis2112 committed:
Merge pull request #13 from noironetworks/ut_ns

added ut for sriov macvlan pod in diff ns

camrossi commented on app/graph.py:
```diff
@@ -471,23 +489,119 @@ def update(self):
+                                for iface_name, link in i["spec"]["aciTopology"].items():
+                                    pods = link.get("pods", [])
+                                    if pods:
+                                        if "sriov" in i["metadata"]["name"]:
```
All instances of `if "sriov" in i["metadata"]["name"]:` should be replaced with `if "sriov" in i["spec"]["primaryCni"]:`
there is a specific field (`primaryCni`) in the `nodefabricnetworkattachments` that can tell us which `primaryCni` we are using and we should not have to rely on the `nodefabricnetworkattachments` name.

burhan20 committed:
differentiated sriov macvlan based on primarycni field

abhis2112 committed:
Merge pull request #14 from noironetworks/primary-cni

differentiated sriov macvlan based on primarycni field

burhan20 committed:
added primarycni in UTs as well

abhis2112 committed:
Merge pull request #15 from noironetworks/ut-changes-primarycni

added primarycni in UTs as well

camrossi commented on helm/templates/nfna_clusterrole.yaml:
```diff
@@ -0,0 +1,29 @@
+  name: "{{ include "vkaci.fullname" . }}-nodefabricnetworkattachments-reader"
+  labels:
+    {{- include "vkaci.labels" . | nindent 4 }}
+  namespace: aci-containers-system
```
This fails in a cluster where CNO is not installed because the `aci-containers-system` does not exists. 
This role should only be created in the CNO case or if we want to create it anyway we should put it in the same ns as vkaci i.e.:

burhan20 committed:
added sec pod iface name based on int value, added pod name to pod graph vertex, added cno false

burhan20 committed:
added cno check for role and role binding

abhis2112 committed:
Merge pull request #16 from noironetworks/pod-iface-name

added sec pod iface name based on int value, added pod name to pod gr…

camrossi commented on helm/templates/clusterrole.yaml:
```diff
@@ -1,4 +1,5 @@
 # Add permission to list the POD for in cluster PODs
+{{- if .Values.cno }}
```
This is not correct, this role is always needed as it gives us the ability to read the PODs as well as other things like calico and cilium CRD that are technically not needed all at the same time but well that is how it was done.

burhan20 committed:
reverted cno check for clusterrole

abhis2112 committed:
Merge pull request #17 from noironetworks/cno

reverted cno check for clusterrole

samiib commented on app/graph.py:
```diff
@@ -214,6 +222,9 @@ def add_neighbour(self, node, neighbour):
                 node['neighbours'][neighbour_adj.sysName]['switches'][switch] = set()
                 logger.info("Found %s as Neighbour to %s:", switch, neighbour_adj.sysName)
+
+            # TODO: Get all the connected ifaces apart from primary and 
```
Is this completed or still needs to be finished?

abhis2112 replied:
this is an older comment will remove

samiib commented on app/graph.py:
```diff
@@ -471,23 +489,119 @@ def update(self):
+                        for key, annotation in annotations.items():
+                            if key == "k8s.v1.cni.cncf.io/network-status":
+                                items_list = json.loads(annotation)
+                                pod_ip = i.status.pod_ip
```
Did you plan to use this local variable? Seems like it's declared but not used.

abhis2112 replied:
noted

samiib commented on app/template/assets/cui-2.0.0/js/vkaci.js:
```diff
@@ -4,10 +4,7 @@ var server_password = "";
 var asnPresent = true;
 
 function neo_viz_config(showPodName, container, cypher, seed = null) {
-    var podCaption = "pod"
```
The tool was designed to show podnames in podview. However, that entire usage has been removed, I think we should at least show podnames where it was shown originally.

abhis2112 replied:
Noted

samiib commented on app/graph.py:
```diff
@@ -471,23 +489,119 @@ def update(self):
+
+        except Exception as e:
+            if e.status == 404:
+                logger.info(f"CRD nodefabricnetworkattachments not detected, SR-IOV/MACVLAN topology support is disabled")
```
Was it meant to be an f string? Or did you guys plan to add a variable to it?

abhis2112 replied:
noted

samiib commented on app/graph.py:
```diff
@@ -441,7 +457,9 @@ def update_node(self, apic, node):
     def update(self):
```
It might be worth refactoring this function into parts because now its quite large and complex.

abhis2112 replied:
Will revisit this in the next iteration for optimization and refactoring

samiib commented on app/template/index.html:
```diff
@@ -134,9 +134,22 @@
                            <button class="btn btn--white btn--dropdown selected view">Different Views</button>
```
The legend and help information needs to be updated.

abhis2112 replied:
Will cover this in the next iteration/PR

samiib reviewed (changes requested):
If the read me can be updated based on the latest changes would be really appreciated.

abhis2112 committed:
Resolving commentds on latest PR
1. https://github.com/datacenter/ACI-Kubernetes-Visualiser/pull/50#discussion_r1320907933
2. https://github.com/datacenter/ACI-Kubernetes-Visualiser/pull/50#discussion_r1320908705
3. https://github.com/datacenter/ACI-Kubernetes-Visualiser/pull/50#discussion_r1322176483
4. https://github.com/datacenter/ACI-Kubernetes-Visualiser/pull/50#discussion_r1322398414

abhis2112 commented:
> If the read me can be updated based on the latest changes would be really appreciated.

Sure, Will look into it
